File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        set<int> s(nums.begin(), nums.end());
        
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        int n = nums.size();
        for(int i=0; i<n-2; i++){
            if (i>0 && nums[i] == nums[i-1]) continue;
            
            int left = i+1;
            int right = n-1;
            
            
            while(left < right){
                 // cout<<i<<" "<<left<<" "<<right<<"\n";
               
                int sum = nums[i] + nums[left] + nums[right];
                if(sum == 0){
                    ans.push_back({nums[i], nums[left], nums[right]});
                    left++;
                    right--;
                    
                    while(left < right && nums[left] == nums[left-1]) left++;
                    while(left < right && nums[right] == nums[right+1]) right--;
                    
                }
                else if(sum > 0){
                    right--;
                } 
                else{
                    left++;
                }
                

            }
        }
        
        return ans;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        ans = []
        for i in range(0, n):
            if i > 0 and nums[i] == nums[i-1]:
                continue
                
            left = i+1
            right = n-1
            
            while left < right :
                sum = nums[i] + nums[left] + nums[right]
                if sum < 0:
                    left+=1
                elif sum > 0:
                    right-=1
                
                else:
                    ans.append([nums[i],nums[left],nums[right]])
                    left+=1
                    right-=1
                    
                    while left < right and nums[left] == nums[left -1]:
                        left+=1
                    
                    while left < right and nums[right] == nums[right+1]:
                        right-=1
        
        return ans
                    
